backend/push.js

```diff
@@ -1,53 +1,292 @@
+  subscriptionInfo.endpoint = new URL(subscriptionInfo.endpoint)
+
+  Object.defineProperties(subscriptionInfo, {
+    'id': {
+      get () {
+        return subcriptionId
+      }
+    },
+    'encryptionKeys': {
+      get: (() => {
+        let count = 0
+        let resultPromise
+        let salt
+        let uaPublic
+
+        return function (this: Object) {
+          // Rotate encryption keys every 2**32 messages
+          // This is just a precaution for a birthday attack, which reduces the
+          // odds of a collision due to salt reuse to under 10**-18.
+          if ((count | 0) === 0) {
+            if (!salt) {
+              // $FlowFixMe[incompatible-call]
+              salt = Buffer.from(this.keys.auth, 'base64url')
+            }
+            if (!uaPublic) {
+              // $FlowFixMe[incompatible-call]
+              uaPublic = Buffer.from(this.keys.p256dh, 'base64url')
+            }
+
+            resultPromise = rfc8291Ikm(uaPublic, salt)
+            count = 1
+          } else {
+            count++
+          }
+
+          return resultPromise
+        }
+      })()
+    },
+    'sockets': {
+      value: new Set()
+    },
+    'subscriptions': {
+      value: new Set(channelIDs)
+    }
+  })
+
+  Object.freeze(subscriptionInfo)
+
+  return subscriptionInfo
+}
 
 export const pushServerActionhandlers: any = {
   [PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY] () {
     const socket = this
-    socket.send(createMessage(REQUEST_TYPE.PUSH_ACTION, publicKey))
+    socket.send(createMessage(REQUEST_TYPE.PUSH_ACTION, { type: PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY, data: getVapidPublicKey() }))
   },
-  [PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION] (payload) {
+  async [PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION] (payload) {
     const socket = this
-    const subscription = JSON.parse(payload)
+    const { server } = socket
+    const subscription = payload
+    const subscriptionId = await getSubscriptionId(subscription)
 
-    // Reference: is it safe to use 'endpoint' as a unique identifier of a push subscription
-    // (https://stackoverflow.com/questions/63767889/is-it-safe-to-use-the-p256dh-or-endpoint-keys-values-of-the-push-notificatio)
-    socket.server.pushSubscriptions[subscription.endpoint] = subscription
+    if (!server.pushSubscriptions[subscriptionId]) {
+      server.pushSubscriptions[subscriptionId] = subscriptionInfoWrapper(subscriptionId, subscription)
+      addSubscriptionToIndex(subscriptionId).then(() => {
+        return sbp('chelonia/db/set', `_private_webpush_${subscriptionId}`, JSON.stringify({ subscription: subscription, channelIDs: [] }))
+      }).catch((e) => console.error(e, 'Error saving subscription', subscriptionId))
+      postEvent(server.pushSubscriptions[subscriptionId], JSON.stringify({ type: 'initial' })).catch(e => console.warn(e, 'Error sending initial push notification'))
```

Comment:
Could you please add a comment explaining what this `'initial'` message is and who/what handles it?